homeassistant/components/tplink/switch.py

```diff
@@ -1,158 +1,109 @@
+    ),
+)
+
+SWITCH_DESCRIPTIONS_MAP = {desc.key: desc for desc in SWITCH_DESCRIPTIONS}
+
+
 async def async_setup_entry(
     hass: HomeAssistant,
-    config_entry: ConfigEntry,
+    config_entry: TPLinkConfigEntry,
     async_add_entities: AddEntitiesCallback,
 ) -> None:
     """Set up switches."""
-    data: TPLinkData = hass.data[DOMAIN][config_entry.entry_id]
+    data = config_entry.runtime_data
     parent_coordinator = data.parent_coordinator
-    device = cast(SmartPlug, parent_coordinator.device)
-    if not device.is_plug and not device.is_strip and not device.is_dimmer:
-        return
-    entities: list = []
-    if device.is_strip:
-        # Historically we only add the children if the device is a strip
-        _LOGGER.debug("Initializing strip with %s sockets", len(device.children))
-        entities.extend(
-            SmartPlugSwitchChild(device, parent_coordinator, child)
-            for child in device.children
-        )
-    elif device.is_plug:
-        entities.append(SmartPlugSwitch(device, parent_coordinator))
+    device = parent_coordinator.device
 
-    # this will be removed on the led is implemented
-    if hasattr(device, "led"):
-        entities.append(SmartPlugLedSwitch(device, parent_coordinator))
+    entities = CoordinatedTPLinkFeatureEntity.entities_for_device_and_its_children(
+        device,
+        coordinator=parent_coordinator,
+        feature_type=Feature.Switch,
+        entity_class=TPLinkSwitch,
+        descriptions=SWITCH_DESCRIPTIONS_MAP,
+    )
 
     async_add_entities(entities)
 
 
-class SmartPlugLedSwitch(CoordinatedTPLinkEntity, SwitchEntity):
-    """Representation of switch for the LED of a TPLink Smart Plug."""
-
-    device: SmartPlug
-
-    _attr_translation_key = "led"
-    _attr_entity_category = EntityCategory.CONFIG
-
-    def __init__(
-        self, device: SmartPlug, coordinator: TPLinkDataUpdateCoordinator
-    ) -> None:
-        """Initialize the LED switch."""
-        super().__init__(device, coordinator)
-        self._attr_unique_id = f"{device.mac}_led"
-        self._async_update_attrs()
-
-    @async_refresh_after
-    async def async_turn_on(self, **kwargs: Any) -> None:
-        """Turn the LED switch on."""
-        await self.device.set_led(True)
-
-    @async_refresh_after
-    async def async_turn_off(self, **kwargs: Any) -> None:
-        """Turn the LED switch off."""
-        await self.device.set_led(False)
-
-    @callback
-    def _async_update_attrs(self) -> None:
-        """Update the entity's attributes."""
-        self._attr_is_on = self.device.led
-
-    @callback
-    def _handle_coordinator_update(self) -> None:
-        """Handle updated data from the coordinator."""
-        self._async_update_attrs()
-        super()._handle_coordinator_update()
-
-
-class SmartPlugSwitch(CoordinatedTPLinkEntity, SwitchEntity):
-    """Representation of a TPLink Smart Plug switch."""
+class TPLinkSwitch(CoordinatedTPLinkFeatureEntity, SwitchEntity):
```

Comment:
```suggestion
class TPLinkSwitchEntity(CoordinatedTPLinkFeatureEntity, SwitchEntity):
```

Reply:
Done in https://github.com/home-assistant/core/pull/120404